kotlin/workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt

```diff
@@ -18,24 +18,41 @@
 package com.squareup.workflow
 
 import com.squareup.workflow.WorkflowAction.Companion.toString
-import com.squareup.workflow.WorkflowAction.Mutator
+import com.squareup.workflow.WorkflowAction.Updater
 
 /**
  * A function that can change the current state of a [Workflow] by returning a new one, and
  * also optionally emit an output.
  */
-interface WorkflowAction<StateT, out OutputT : Any> {
+interface WorkflowAction<StateT, OutputT : Any> {
+  @Deprecated("Use Updater")
   class Mutator<S>(var state: S)
 
-  /**
-   * Called with the current state of a workflow, allowing the state to be updated.
-   *
-   * @return an output value for the workflow to emit, or null to emit nothing.
-   */
-  fun Mutator<StateT>.apply(): OutputT?
+  class Updater<S, O : Any>(var nextState: S) {
+    private var _nextOutput: O? = null
 
-  companion object {
+    var nextOutput: O?
+      get() = _nextOutput
+      set(value) {
```

Comment:
I'm not convinced that this is a good idea. It requires a lot of this kind of thing in order to avoid runtime exceptions:

```
      output?.let { nextOutput = it }
```

e.g. see line 199 of the Dungeon app's `GameWorkflow`, where I had to debug a crash.

Reply:
I don't think it's a good idea either, I was thinking this parameter would be non-nullable (and it can throw if you try reading before writing). In that case, maybe this should actually be a write-only method instead of a var.

Reply:
In particular, relying on nullability here is going to make it harder to remove the non-nullable bound on `OutputT` in the future.

Reply:
I like the setter-method-only notion. We don't owe anyone an `output` field.

Reply:
```kotlin
    /**
     * Sets the value workflow will emit when this action is applied. If this method
     * is not called, there will be no output.
     */
    fun setOutput(output: O)
```